### PR TITLE
[ISSUE #1869] Delay message can't be consumed when delay offset in delayOffset.json is wrong

### LINKED ISSUE #1869: delay message can't be consumed
**BUG REPORT**

1. Please describe the issue you observed:

- What did you do (The steps to reproduce)?
In our production environment，producer send delay message，but this message can not be consumed.

Producer code 
Message msg = new Message("my_test_topic",
value.getBytes(RemotingHelper.DEFAULT_CHARSET));
msg.setKeys(key);
msg.setDelayTimeLevel(4);
SendResult sendResult = producer.send(msg);

- What did you expect to see?

This message can be consumed correct.

- What did you see instead?
In the storeerror.log there are a lot of error message.
2020-03-18 16:38:06 WARN ScheduleMessageTimerThread - Offset not matched. Request offset: 509548460, firstOffset: 0, lastOffset: 6000000, mappedFileSize: 6000000, mappedFiles count: 1
2020-03-18 16:38:06 WARN ScheduleMessageTimerThread - Offset not matched. Request offset: 665682360, firstOffset: 0, lastOffset: 6000000, mappedFileSize: 6000000, mappedFiles count: 1
2020-03-18 16:38:07 WARN ScheduleMessageTimerThread - Offset not matched. Request offset: 509548460, firstOffset: 0, lastOffset: 6000000, mappedFileSize: 6000000, mappedFiles count: 1
2020-03-18 16:38:07 WARN ScheduleMessageTimerThread - Offset not matched. Request offset: 665682360, firstOffset: 0, lastOffset: 6000000, mappedFileSize: 6000000, mappedFiles count: 1
2020-03-18 16:38:07 WARN ScheduleMessageTimerThread - Offset not matched. Request offset: 509548460, firstOffset: 0, lastOffset: 6000000, mappedFileSize: 6000000, mappedFiles count: 1
2020-03-18 16:38:07 WARN ScheduleMessageTimerThread - Offset not matched. Request offset: 665682360, firstOffset: 0, lastOffset: 6000000, mappedFileSize: 6000000, mappedFiles count: 1
2020-03-18 16:38:07 WARN ScheduleMessageTimerThread - Offset not matched. Request offset: 509548460, firstOffset: 0, lastOffset: 6000000, mappedFileSize: 6000000, mappedFiles count: 1
2020-03-18 16:38:07 WARN ScheduleMessageTimerThread - Offset not matched. Request offset: 665682360, firstOffset: 0, lastOffset: 6000000, mappedFileSize: 6000000, mappedFiles count: 1

2. Please tell us about your environment:
Our production environment. RocketMQ 4.5.2；But other rocketmq cluster of this version run ok.

3. Other information (e.g. detailed explanation, logs, related issues, suggestions how to fix, etc):
After remove file "consumequeue/SCHEDULE_TOPIC_XXXX" and remove "config/delayOffset.json" file, restart the broker,  delay message restore, consume correctly. 

Reference event review process：
https://mp.weixin.qq.com/s/f27tcOlm0_BTvL8725RPTQ

### PR DESCRIPTION
**Pull request by RongtongJin**
**Make sure set the target branch to `develop`**

## What is the purpose of the change

Fix #1869
Delay message can't be consumed when delay offset in delayOffset.json is wrong

## Brief changelog

Correct delay offset when ScheduleMessageService init and timer task running.

## Verifying this change

XXXX

Follow this checklist to help us incorporate your contribution quickly and easily. Notice, `it would be helpful if you could finish the following 5 checklist(the last one is not necessary)before request the community to review your PR`.

- [x] Make sure there is a [Github issue](https://github.com/apache/rocketmq/issues) filed for the change (usually before you start working on it). Trivial changes like typos do not require a Github issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue. 
- [x] Format the pull request title like `[ISSUE #123] Fix UnknownException when host config not exist`. Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit-test(over 80% coverage) to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/apache/rocketmq/tree/master/test).
- [x] Run `mvn -B clean apache-rat:check findbugs:findbugs checkstyle:checkstyle` to make sure basic checks pass. Run `mvn clean install -DskipITs` to make sure unit-test pass. Run `mvn clean test-compile failsafe:integration-test`  to make sure integration-test pass.
- [ ] If this contribution is large, please file an [Apache Individual Contributor License Agreement](http://www.apache.org/licenses/#clas).

### CODE REVIEW

**Comment by RongtongJin** [comment]
PTAL @yongliangcode

**Review by panzhi33 — APPROVED** [review]

**Review by duhenglucky — APPROVED** [review]

File: store/src/main/java/org/apache/rocketmq/store/schedule/ScheduleMessageService.java [hunk 1]
```diff
@@ -376,10 +408,17 @@ public void executeOnTimeup() {
...
-                        log.error("schedule CQ offset invalid. offset=" + offset + ", cqMinOffset="
-                            + cqMinOffset + ", queueId=" + cq.getQueueId());
+                        log.error("schedule CQ offset invalid. offset={}, cqMinOffset={}, cqMaxOffset={}, queueId={}",
+                            offset, cqMinOffset, cqMaxOffset, cq.getQueueId());
+                    }
+
+                    if (offset > cqMaxOffset) {
+                        failScheduleOffset = cqMaxOffset;
+                        log.error("schedule CQ offset invalid. offset={}, cqMinOffset={}, cqMaxOffset={}, queueId={}",
+                            offset, cqMinOffset, cqMaxOffset, cq.getQueueId());
```
Comment (ShannonDing) [inline store/src/main/java/org/apache/rocketmq/store/schedule/ScheduleMessageService.java:421]: it is better to print the “failScheduleOffset” value at last.

**Review by guyinyou — APPROVED** [review]

File: store/src/main/java/org/apache/rocketmq/store/schedule/ScheduleMessageService.java (hunk 1 above)
Comment (RongtongJin) [reply store/src/main/java/org/apache/rocketmq/store/schedule/ScheduleMessageService.java:421]: When offset > cqMaxOffset  failScheduleOffset = cqMaxOffset, when offset < cqMinOffset  failScheduleOffset = cqMinOffset. In general, failscheduleoffset = offset. So all relevant offset values have been printed in the current log. There is no need to print failScheduleOffset at last.

File: store/src/main/java/org/apache/rocketmq/store/schedule/ScheduleMessageService.java [hunk 2]
```diff
@@ -133,7 +132,8 @@ public void start() {
                 @Override
                 public void run() {
                     try {
-                        if (started.get()) ScheduleMessageService.this.persist();
+                        if (started.get())
+                            ScheduleMessageService.this.persist();
```
Comment (ltamber) [inline store/src/main/java/org/apache/rocketmq/store/schedule/ScheduleMessageService.java:136]: is it better to surrounded by `{}` ?

**Review by maixiaohai — APPROVED** [review]

File: store/src/main/java/org/apache/rocketmq/store/schedule/ScheduleMessageService.java [hunk 3]
```diff
@@ -165,9 +166,46 @@ public String encode() {
...
+                long cqMaxOffset = cq.getMaxOffsetInQueue();
+                if (currentDelayOffset < cqMinOffset) {
+                    correctDelayOffset = cqMinOffset;
+                    log.error("schedule CQ offset invalid. offset={}, cqMinOffset={}, cqMaxOffset={}, queueId={}",
+                        currentDelayOffset, cqMinOffset, cqMaxOffset, cq.getQueueId());
+                }
+
+                if (currentDelayOffset > cqMaxOffset) {
+                    correctDelayOffset = cqMaxOffset;
+                    log.error("schedule CQ offset invalid. offset={}, cqMinOffset={}, cqMaxOffset={}, queueId={}",
```
Comment (vmstar) [inline store/src/main/java/org/apache/rocketmq/store/schedule/ScheduleMessageService.java:194]: currentDelayOffset < cqMinOffset 和 currentDelayOffset > cqMaxOffset 两个场景对correctDelayOffset进行调整，日志上是不是要有个区别。

Comment (duhenglucky) [reply store/src/main/java/org/apache/rocketmq/store/schedule/ScheduleMessageService.java:194]: @vmstar Would you like submit a PR to polish it？

**Comment by woaishixiaoxiao** [comment]
当过期的文件被删除的时候会出现offset < minoffset的情况
那什么时候 会出现 offset > maxoffset的情况呢？

**Pull request merged**